portal/account/account.lib.php

```diff
@@ -152,42 +149,50 @@ function doCredentials($pid)
     global $srcdir;
     require_once("$srcdir/authentication/common_operations.php");
 
-    $newpd = sqlQuery("SELECT * FROM `patient_data` WHERE `pid`=?", array(
+    $newpd = sqlQuery("SELECT id,fname,mname,lname,email,email_direct FROM `patient_data` WHERE `pid`=?", array(
         $pid
     ));
 
+    $uname = $newpd['fname'] . $newpd['id'];
+    // Token expiry 1 hour
+    $expiry = new DateTime('NOW');
+    $expiry->add(new DateInterval('PT01H'));
+
     $clear_pass = generatePassword();
 
-    $uname = $newpd['fname'] . $newpd['id'];
+    $token = RandomGenUtils::createUniqueToken(32);
+    $one_time = hash('sha256', $token);
 
-    $res = sqlStatement("SELECT * FROM patient_access_onsite WHERE pid=?", array(
-        $pid
-    ));
-    $query_parameters = array(
-        $uname
-    );
-    $salt_clause = "";
-    // For onsite portal create a blowfish based hash and salt.
+    $encoded_link = sprintf("%s?%s", attr($GLOBALS['portal_onsite_two_address']), http_build_query([
+        'forward' => bin2hex($token),
+        'validate' => bin2hex($expiry->format('U')),
+        'site' => $_SESSION['site_id']
+    ]));
+
+    $res = sqlStatement("SELECT * FROM patient_access_onsite WHERE pid=?", array($pid));
+    $query_parameters = array($uname,$one_time);
     $new_salt = oemr_password_salt();
     $salt_clause = ",portal_salt=? ";
     array_push($query_parameters, oemr_password_hash($clear_pass, $new_salt), $new_salt);
     array_push($query_parameters, $pid);
     if (sqlNumRows($res)) {
-        sqlStatement("UPDATE patient_access_onsite SET portal_username=?,portal_pwd=?,portal_pwd_status=0 " . $salt_clause . " WHERE pid=?", $query_parameters);
+        sqlStatement("UPDATE patient_access_onsite SET portal_username=?,portal_login_username=?,portal_pwd=?,portal_pwd_status=0 " . $salt_clause . " WHERE pid=?", $query_parameters);
     } else {
-        sqlStatement("INSERT INTO patient_access_onsite SET portal_username=?,portal_pwd=?,portal_pwd_status=0" . $salt_clause . " ,pid=?", $query_parameters);
+        sqlStatement("INSERT INTO patient_access_onsite SET portal_username=?,portal_login_username=?,portal_pwd=?,portal_pwd_status=0" . $salt_clause . " ,pid=?", $query_parameters);
```

Comment:
for me to figure out later in code review:
why $one_time is stored in portal_login_username

Reply:
It hides the new one time temp pass which is already encrypted and encoded. Currently valid for only an hour. Also insures username is reset as this is for a password reset after all.
Patient can keep same username but I force a reentry.
A portal reset from demographics will automatic clear login username

Reply:
Thanks for the explanation.